vpn/dylib/lib.go

```diff
@@ -0,0 +1,93 @@
+//go:build darwin
+// +build darwin
+
+package main
+
+import "C"
+
+import (
+	"context"
+	"io"
+	"os"
+
+	"golang.org/x/sys/unix"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+
+	"github.com/coder/coder/v2/vpn"
+)
+
+// OpenTunnel creates a new VPN tunnel by `dup`ing the provided 'PIPE'
+// file descriptors for reading, writing, and logging.
+//
+//export OpenTunnel
+func OpenTunnel(cReadFD, cWriteFD int32) int32 {
+	ctx := context.Background()
+
+	conn, err := newFdConn(cReadFD, cWriteFD)
+	if err != nil {
+		return -1
+	}
+
+	// Logs will be sent over the protocol
+	_, err = vpn.NewTunnel(ctx, slog.Make(sloghuman.Sink(io.Discard)), conn)
+	if err != nil {
+		return -1
+	}
+
+	return 0
+}
+
+type pipeConn struct {
+	r *os.File
+	w *os.File
+}
+
+func (f *pipeConn) Read(p []byte) (n int, err error) {
+	return f.r.Read(p)
+}
+
+func (f *pipeConn) Write(p []byte) (n int, err error) {
+	return f.w.Write(p)
+}
+
+func (f *pipeConn) Close() error {
+	rErr := f.r.Close()
+	wErr := f.w.Close()
+	if rErr != nil {
+		return rErr
+	}
+	return wErr
+}
+
+func newFdConn(cReadFD, cWriteFD int32) (io.ReadWriteCloser, error) {
+	readFD, err := unix.Dup(int(cReadFD))
+	if err != nil {
+		return nil, xerrors.Errorf("dup readFD: %w", err)
+	}
+	reader := os.NewFile(uintptr(readFD), "PIPE")
+	if reader == nil {
+		unix.Close(readFD)
```

Comment:
This condition can only be met if the FD is invalid anyways, so this close call does nothing